### ISSUE TITLE
is there a way to know whether ack/nack succeeds or fails

### ISSUE THREAD

**Issue by shohi** [apache/pulsar-client-go#296]
From the source code, no callback or returned error is found.

```go
type Consumer interface {
        // Ack the consumption of a single message
	Ack(Message)
       
        // Acknowledge the failure to process a single message.
	//
	// When a message is "negatively acked" it will be marked for redelivery after
	// some fixed delay. The delay is configurable when constructing the consumer
	// with ConsumerOptions.NAckRedeliveryDelay .
	//
	// This call is not blocking.
	Nack(Message)

       // ...
}




```

**Comment by sijie** [comment]
@shohi Currently the protocol of acknowledge doesn't support any response. It is fire-and-forget. We will add responses for acknowledgment as part of transaction support.

**Comment by shohi** [comment]
@sijie thanks for your reply, glad to hear it will come along with transaction support. And I also wonder if it's possible to provide callback handlers for connection/operation errors, thus users can utilize them to do some logging or monitoring.

**Comment by merlimat** [comment]
> And I also wonder if it's possible to provide callback handlers for connection/operation errors, thus users can utilize them to do some logging or monitoring.

The connection aspect are hidden by design, to provide higher-level API components only. The internal reconnections are already logged.

**Comment by shohi** [comment]
> > And I also wonder if it's possible to provide callback handlers for connection/operation errors, thus users can utilize them to do some logging or monitoring.
> 
> The connection aspect are hidden by design, to provide higher-level API components only. The internal reconnections are already logged.

Yeah, but IMHO it's useful to notify users that exceptions/unrecoverable errors have occurred. In practice, we want to monitor the out connections and log key events with user-defined error/event codes.

**Comment by shohi** [comment]
It will be even better that Ack/Nack could support timeout or context. There is a related issue of Ack timeout - https://github.com/apache/pulsar-client-go/issues/217, i think the ack timeout is helpful in some scenarios where the client only consumes one specific topic.
